File: py/Languages/LanguageSelector.py

import os
from mimetypes import MimeTypes

from py.Languages.PythonLanguage import PythonLanguage
from py.Languages.PHPLanguage import PHPLanguage
from py.Languages.MarkdownLanguage import MarkdownLanguage
from py.Hub import Log, Hub

class LanguageSelector:

    def __init__(self, hub: Hub):
        self.mime = MimeTypes()
        self.hub = hub
    
    def selectForFilePath(self, filePath):
        language = None
    
        (mimeType, n) = self.mime.guess_type(filePath)
        if mimeType != None:
            language = self.selectForMimeType(mimeType)
            
        fileExtension = os.path.splitext(filePath)[-1]
        if fileExtension == ".md":
            language = MarkdownLanguage(self.hub)
        
        Log.debug("Selected " + str(type(language)) + " for " + filePath)
        
        return language
        
    def selectForMimeType(self, mimeType):
        if mimeType == "text/x-python":
            return PythonLanguage(self.hub)
            
        if mimeType == "application/x-python-code":
            return None # cached / compiled python bytecode
        
        if mimeType == "text/php":
            return PHPLanguage(self.hub)
        
        print(mimeType)
        
        [category, subcategory] = mimeType.split("/")
        print(category)
        print(subcategory)

File: java/src/main/java/de/addiks/einsicht/languages/LanguageSelector.java
package de.addiks.einsicht.languages;

import de.addiks.einsicht.Hub;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

public class LanguageSelector {
    private static final Logger log = LoggerFactory.getLogger(LanguageSelector.class);
    private final Hub hub;

    public LanguageSelector(Hub hub) {
        this.hub = hub;
    }

    public Language selectForFilePath(Path filePath) throws IOException {
        Language language = null;

        String mimeType = Files.probeContentType(filePath);
        if (mimeType != null) {
            if (mimeType.equals("text/x-python")) {
                language = new PythonLanguage(hub);

            } else if (mimeType.equals("application/x-python-code")) {
                // Python bytecode

            } else if (mimeType.equals("text/php")) {

            }

            if (language == null) {
                log.warn("Unknown mime-type '{}'!", mimeType);
            }
        }

        if (language == null) {
            if (filePath.endsWith(".md")) {

            } else if (filePath.endsWith(".java")) {
                language = new JavaLanguage(hub);
            }
        }

        if (language == null) {
            log.warn("Could not determine lanugage for file '{}'!", filePath);
            language = new UnknownLanguage(hub);
        }

        log.info("Determined language of '{}' to be {}.", filePath, language.name());
        return language;
    }


}
